### PR TITLE
feature[TW27514]: Update relation subquery to improve performance

### PR DESCRIPTION
**Pull request by audrey-denk**
Change-Id: I9a23c80a69b4d60d81dd56ac89c988351b1202f9

### CODE REVIEW

**Review by ryanbase — APPROVED** [review]

File: plugins/org.eclipse.osee.orcs.db/src/org/eclipse/osee/orcs/db/internal/sql/SelectiveArtifactSqlWriter.java [hunk 1]
```diff
@@ -356,56 +356,18 @@ private void writeFieldsCommonTableExpression(String artWithAlias, String attsAl
...
-         }
-
-         @Override
-         public int getPriority() {
-            return SqlHandlerPriority.RELATED_TO_ART_IDS.ordinal();
-         }
-      });
-      relWriter.write(handlers);
-      write(relWriter.toSql());
+      writeSelectAndHint();
```
Comment (lucianovaglienti) [inline plugins/org.eclipse.osee.orcs.db/src/org/eclipse/osee/orcs/db/internal/sql/SelectiveArtifactSqlWriter.java:362]: Q: wouldn't this cause  if (relsAlias == null && rels2Alias == null) {
         writeSelectFields(artAlias, "art_id", artAlias, "art_type_id", txAlias, "app_id", txAlias, "transaction_id",
            txAlias, "mod_type", txAlias, "tx_current", txAlias, "gamma_id as art_gamma_id");
         if (OptionsUtil.getIncludeApplicabilityTokens(rootQueryData.getOptions())) {
            writeSelectFields(getMainTableAlias(OseeDb.OSEE_KEY_VALUE_TABLE), "value app_value");
         } to get hit multiple times? Is that intentional?

File: plugins/org.eclipse.osee.orcs.db/src/org/eclipse/osee/orcs/db/internal/sql/SelectiveArtifactSqlWriter.java [hunk 2]
```diff
@@ -457,11 +420,59 @@ private void writeRelsCommonTableExpression(String artWithAlias) {
       String primary = "primaryArt";
       String secondary = "secondaryArt";
       String txdAlias = "txd";
+      String tuple2Alias = "t2";
+      String tuple2TxsAlias = "t2Txs";
+      writeUseNlTableHint(relAlias + " " + relTxsAlias);
+      writeSelectFields(primary, "*", relAlias, "rel_link_type_id AS type_id");
+      write(
+         ", 'B'  AS value, '' AS spare1, 0 AS spare2 ,%s.gamma_id, b_art_id AS other_art_id, %s.art_type_id as other_art_type_id, %s.gamma_id as other_art_gamma_id, 0 as rel_type, 0 as rel_order",
```
Comment (lucianovaglienti) [inline plugins/org.eclipse.osee.orcs.db/src/org/eclipse/osee/orcs/db/internal/sql/SelectiveArtifactSqlWriter.java:428]: Just a thought...can this go in writeSelectFields somehow? or can we make the 0 as rel_type, 0 as rel_order be prepared variables

Comment (lucianovaglienti) [inline plugins/org.eclipse.osee.orcs.db/src/org/eclipse/osee/orcs/db/internal/sql/SelectiveArtifactSqlWriter.java:428]: Can we make sure to leave a space at the end of the string, I don't know if writeSelectFields handles it below

File: plugins/org.eclipse.osee.orcs.db/src/org/eclipse/osee/orcs/db/internal/sql/SelectiveArtifactSqlWriter.java [hunk 3]
```diff
@@ -457,11 +420,59 @@ private void writeRelsCommonTableExpression(String artWithAlias) {
...
+         relTxsAlias,  secondary, secondaryTxsAlias);
```
Comment (lucianovaglienti) [inline plugins/org.eclipse.osee.orcs.db/src/org/eclipse/osee/orcs/db/internal/sql/SelectiveArtifactSqlWriter.java:429]: Maybe a good chunk of this we could turn into a function and reuse below?

File: plugins/org.eclipse.osee.orcs.db/src/org/eclipse/osee/orcs/db/internal/sql/SelectiveArtifactSqlWriter.java [hunk 4]
```diff
@@ -499,11 +522,13 @@ private void writeRelsCommonTableExpression2(String artWithAlias) {
...
       String primary = "primaryArt";
       String secondary = "secondaryArt";
+      String tuple2Alias = "t2";
+      String tuple2TxsAlias = "t2Txs";
       writeUseNlTableHint(relAlias + " " + relTxsAlias);
       writeSelectFields(primary, "*", relAlias, "rel_type AS type_id");
       write(
-         ", CASE %s.art_id WHEN a_art_id THEN 'B' ELSE 'A' END AS value, '' AS spare1, 0 AS spare2, %s.gamma_id, CASE %s.art_id WHEN a_art_id THEN b_art_id ELSE a_art_id END AS other_art_id,%s.gamma_id as other_art_gamma_id, rel_type, rel_order",
-         primary, relTxsAlias, primary, secondaryTxsAlias);
+         ", 'B' value, '' AS spare1, 0 AS spare2, %s.gamma_id, b_art_id AS other_art_id, %s.art_type_id as other_art_type_id, %s.gamma_id as other_art_gamma_id, rel_type, rel_order",
```
Comment (lucianovaglienti) [inline plugins/org.eclipse.osee.orcs.db/src/org/eclipse/osee/orcs/db/internal/sql/SelectiveArtifactSqlWriter.java:530]: Leave a space

File: plugins/org.eclipse.osee.orcs.db/src/org/eclipse/osee/orcs/db/internal/sql/SelectiveArtifactSqlWriter.java (hunk 1 above)
Comment (audrey-denk) [reply plugins/org.eclipse.osee.orcs.db/src/org/eclipse/osee/orcs/db/internal/sql/SelectiveArtifactSqlWriter.java:362]: hmmm... those don't get hit at all actually.

File: plugins/org.eclipse.osee.orcs.db/src/org/eclipse/osee/orcs/db/internal/sql/SelectiveArtifactSqlWriter.java (hunk 2 above)
Comment (audrey-denk) [reply plugins/org.eclipse.osee.orcs.db/src/org/eclipse/osee/orcs/db/internal/sql/SelectiveArtifactSqlWriter.java:428]: Looks like it does...but I can add anyways

File: plugins/org.eclipse.osee.orcs.db/src/org/eclipse/osee/orcs/db/internal/sql/SelectiveArtifactSqlWriter.java (hunk 4 above)
Comment (audrey-denk) [reply plugins/org.eclipse.osee.orcs.db/src/org/eclipse/osee/orcs/db/internal/sql/SelectiveArtifactSqlWriter.java:530]: got it

File: plugins/org.eclipse.osee.orcs.db/src/org/eclipse/osee/orcs/db/internal/sql/SelectiveArtifactSqlWriter.java (hunk 3 above)
Comment (audrey-denk) [reply plugins/org.eclipse.osee.orcs.db/src/org/eclipse/osee/orcs/db/internal/sql/SelectiveArtifactSqlWriter.java]: Done!

**Review by lucianovaglienti — APPROVED** [review]

**Pull request merged**